### PR TITLE
Add typed asOfJoin, delegate aj/raj

### PR DESCRIPTION
**Pull request by devinrsmith**
Partial #3692

### CODE REVIEW

**Comment by rcaudy** [comment]
Per our other discussion, we want to reverse the directionality of match rules to better express what happens. That is, we want aj to use GT and GEQ, raj to use LT and LEQ.

File: java-client/session/src/main/java/io/deephaven/client/impl/BatchTableRequestBuilder.java [hunk 1]
```diff
@@ -390,15 +390,17 @@ public void visit(AsOfJoinTable aj) {
         }
 
         private static AsOfJoinTablesRequest.MatchRule adapt(AsOfJoinRule rule) {
+            // todo: fix gRPC?
```
Comment (rcaudy) [inline java-client/session/src/main/java/io/deephaven/client/impl/BatchTableRequestBuilder.java:393]: We'll update gRPC in another PR before this goes into a release.

File: table-api/src/main/java/io/deephaven/api/expression/AsOfJoinMatchFactory.java [hunk 2]
```diff
@@ -109,55 +92,39 @@ public T getExpression(String expression, Matcher matcher, Object... args) {
...
-            }
-        });
     }
 
-    private static AsOfJoinResult getAjExpression(String match) {
+    private static AsOfJoinMatch getAjExpression(String match) {
         return asOfJoinParser.parse(match);
     }
 
     public static AsOfJoinResult getAjExpressions(String... matches) {
```
[the remark this answers is not in the document]
Comment (devinrsmith) [reply table-api/src/main/java/io/deephaven/api/expression/AsOfJoinMatchFactory.java]: Good idea.

File: java-client/session/src/main/java/io/deephaven/client/impl/BatchTableRequestBuilder.java (hunk 1 above)
Comment (devinrsmith) [reply java-client/session/src/main/java/io/deephaven/client/impl/BatchTableRequestBuilder.java]: yep

File: table-api/src/main/java/io/deephaven/api/AsOfJoinMatch.java [hunk 3]
```diff
@@ -7,18 +7,89 @@
...
             final ColumnName leftColumn,
             final AsOfJoinRule joinRule,
             final ColumnName rightColumn) {
         return ImmutableAsOfJoinMatch.of(leftColumn, joinRule, rightColumn);
     }
 
+    /**
+     * Parses the expression {@code x}, expecting it to either be a single column name, or an expression of the form
+     * {@code "lhsColumnName >= lhsColumnName"} or {@code "lhsColumnName > rhsColumnName"}. When it is a single column
+     * name, a {@link AsOfJoinRule#GREATER_THAN_EQUAL} is used.
```
Comment (rcaudy) [inline table-api/src/main/java/io/deephaven/api/AsOfJoinMatch.java:35]:
```suggestion
     * Parses the expression {@code x}, expecting it to either be a single column name, or an expression of the form
     * {@code "lhsColumnName >= rhsColumnName"} or {@code "lhsColumnName > rhsColumnName"}. When it is a single column
     * name, match rule {@link AsOfJoinRule#GREATER_THAN_EQUAL} is used.
```

File: table-api/src/main/java/io/deephaven/api/AsOfJoinMatch.java [hunk 4]
```diff
@@ -7,18 +7,89 @@
...
+            case GREATER_THAN_EQUAL:
+                return parsed;
+        }
+        throw new IllegalArgumentException("Unsupported operation for aj: " + parsed.joinRule());
+    }
+
+    /**
+     * Parses the expression {@code x}, expecting it to either be a single column name, or an expression of the form
+     * {@code "lhsColumnName <= lhsColumnName"} or {@code "lhsColumnName < rhsColumnName"}. When it is a single column
+     * name, a {@link AsOfJoinRule#LESS_THAN_EQUAL} is used.
```
Comment (rcaudy) [inline table-api/src/main/java/io/deephaven/api/AsOfJoinMatch.java:59]:
```suggestion
     * Parses the expression {@code x}, expecting it to either be a single column name, or an expression of the form
     * {@code "lhsColumnName <= rhsColumnName"} or {@code "lhsColumnName < rhsColumnName"}. When it is a single column
     * name, match rule {@link AsOfJoinRule#LESS_THAN_EQUAL} is used.
```

**Review by rcaudy — APPROVED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

How-to: https://github.com/deephaven/deephaven.io/issues/2694
Conceptual: https://github.com/deephaven/deephaven.io/issues/2693
Reference: https://github.com/deephaven/deephaven.io/issues/2695
Blog: https://github.com/deephaven/deephaven.io/issues/2696

**Pull request merged**